File: Devices/ECG/PatientWrapper.cpp
#include "PatientWrapper.hpp"
#include <vector>
#include <ctime>

void addConsumable(const string& consumable)
{
	temporary_patient.addConsumable(consumable);
}


void addReportId(const string& report_id)
{
	temporary_patient.addReportId(report_id);
}

int isPatientFree()
{
	return !original_patient.getBusyStatus();
}

bool isPatientDataAvailable() {
	if (original_patient.isEmpty()) {
		return false;
	}
	return true;
}

string generate_patient_id(const string& name)
{
	srand((unsigned)time(0));
	int num = rand() % 2000;
	return name + to_string(num);
}


PatientInfo getPatientInfo()
{
	string name;
	int age;
	string gender;
	string procedureName;

	PatientInfo p_info;

	cout << "Enter patient name: ";
	cin >> p_info.name;
	
	cout << "Enter patient gender (Male/Female): ";
	cin >> p_info.gender;
	
	cout << "Enter patient age: ";
	cin >> p_info.age;

	cout << "Enter Procedure name: ";
	cin >> p_info.procedureName;
	
	return p_info;
}

void generateAndsetPatientId()
{
	string name = temporary_patient.getName();
	string patient_id = generate_patient_id(name);
	temporary_patient.setId(patient_id);
}

void setPatientInfo(PatientInfo& p_info)
{
	temporary_patient.setName(p_info.name);
	temporary_patient.setAge(p_info.age);
	temporary_patient.setGender(p_info.gender);
	temporary_patient.setProcedureName(p_info.procedureName);
	generateAndsetPatientId();
}

void addNewPatient()
{
    string strInput;
    string verified = "no";
	int intInput;
	PatientInfo p_info;
	do {
		p_info = getPatientInfo();
		cout << "Verified your details (yes/no)? ";
		cin >> verified;
	} while (verified == "no");

	setPatientInfo(p_info);

    serialized_data = temporary_patient.toString();
	publishable = true;
}

void setName(const string& name)
{
	temporary_patient.setName(name);
}

void setAge(int intInput)
{
	temporary_patient.setAge(intInput);
}

void setGender(const string& strInput)
{
	temporary_patient.setGender(strInput);
}

void setProcedureName(const string& strInput)
{
	temporary_patient.setProcedureName(strInput);
}

void editName()
{
	string strInput;
	cout << "Enter new Name: ";
	cin >> strInput;
	setName(strInput);
}

void editAge()
{
	int intInput;
	cout << "Enter new age: ";
	cin >> intInput;
	setAge(intInput);
}

void editGender()
{
	string strInput;
	cout << "Enter new gender: ";
	cin >> strInput;
	setGender(strInput);	
}

void editProcedure()
{
	string strInput;
	cout << "Enter new procedure name: ";
	cin >> strInput;
	setProcedureName(strInput);
}


void editAll()
{
	editName();
	editAge();
	editGender();
	editProcedure();
}

void setAndPublishPatientBusy()
{
	temporary_patient.setBusyStatus(1);
	serialized_data = temporary_patient.toString();
	publishable = true;
	PublishToSynchronize();
}

void unsetPatientBusyStatus()
{
	temporary_patient.setBusyStatus(0);
}

void editPatientInfo()
{
	if (!isPatientDataAvailable()) {
		cout << "Sorry.. no patient data is available. Kindly add patient info before editing..." << endl;
		return;
	}

	if(!isPatientFree()) {
		cout << "The patient data is being used by some other device. Try again after sometime..." << endl;
		return;
	}


	setAndPublishPatientBusy();


	cout << "The following are the entered patient information: " << endl;
	printPatientInformation();

	displayEditMenu();
	
	int choice = getUserChoice();
	ExecuteEditMenuChoice(choice);

	unsetPatientBusyStatus();
	
	serialized_data = temporary_patient.toString();
	publishable = true;
}

void printPatientInformation()
{
    cout << "Patient Name: " << original_patient.getName() << endl;
	cout << "Patent Gender: " << original_patient.getGender() << endl;
	cout << "Patient Age: " << original_patient.getAge() << endl;
	cout << "Procedure Name: " << original_patient.getProcedureName() << endl;
	cout << "Consumables: " << original_patient.getConsumables() << endl;
	cout << "ReportIds: " << original_patient.getReportIds() << endl;
}

File: Devices/PythonClient/PatientWrapper.py
import settings
import time
import random
from MenuCLI import MenuCLI

class PatientWrapper:
    @staticmethod
    def setAndPublishBusyStatus(client, busyStatus):
        settings.temporary_patient.setBusyStatus(busyStatus)
        msg = settings.temporary_patient.toString()
        client.publish(settings.topic, msg)


    @staticmethod
    def isPatientBusy():
        return settings.original_patient.getBusyStatus()


    @staticmethod
    def isPatientDataAvailable():
        return settings.original_patient.getId != 0


    @staticmethod
    def EditName():
        name = input("Enter corrected name: ")
        settings.temporary_patient.setName(name)


    @staticmethod
    def EditAge():
        age = int(input("Enter corrected age: "))
        settings.temporary_patient.setAge(age)

    @staticmethod
    def EditGender():
        gender = input("Enter corrected gender: ")
        settings.temporary_patient.setGender(gender)

    @staticmethod
    def EditProcedureName():
        procedureName = input("Enter procedure name: ")
        settings.temporary_patient.setProcedureName(procedureName)

    @staticmethod
    def EditAll():
        PatientWrapper.EditName()
        PatientWrapper.EditAge()
        PatientWrapper.EditGender()
        PatientWrapper.EditProcedureName()

    @staticmethod
    def isPatientDataAvailableAndFree():
        if not PatientWrapper.isPatientDataAvailable():
            print("No patient details entered.. Please try after entering the patient details!")
            return False
        
        if PatientWrapper.isPatientBusy():
            print("Sorry! some procedure is already in progress with the current patient.. please try again in sometime..") 
            return False
        
        return True
    
    @staticmethod 
    def procedure(repeat_count):
        count = 0
        while count < repeat_count:
            time.sleep(1)
            strInput = input("Enter any consumable used till now: (if nothing, type none) ")
            if strInput != "none":
                settings.temporary_patient.addConsumables(strInput)
            count += 1

    @staticmethod
    def startProcedure():
        client = settings.client

        if not PatientWrapper.isPatientDataAvailableAndFree:
            return
        
        PatientWrapper.setAndPublishBusyStatus(client, 1)

        PatientWrapper.procedure(repeat_count = 3)
        reportId = "MRI " + str(random.randint(100, 3000))
        settings.temporary_patient.addReportIds(reportId)

    
        settings.temporary_patient.setBusyStatus(0)
        msg = settings.temporary_patient.toString()
        client.publish(settings.topic, msg)


    @staticmethod
    def get_serialized_patient_info():
        name = input("Enter patient name: ")
        age = int(input("Enter age: "))
        gender = input("Enter gender: ")
        procedure_name = input("Enter procedure name: ")

        settings.temporary_patient.setId(settings.get_patient_count())
        settings.temporary_patient.setName(name)
        settings.temporary_patient.setAge(age)
        settings.temporary_patient.setGender(gender)
        settings.temporary_patient.setProcedureName(procedure_name)

        return settings.temporary_patient.toString()

    @staticmethod
    def addNewPatient():
        client = settings.client
        settings.update_patient_count()
        msg = PatientWrapper.get_serialized_patient_info()
        client.publish(settings.topic, msg)
        while not client.published:
            time.sleep(1)


    @staticmethod
    def editPatient():
        client = settings.client
        if not PatientWrapper.isPatientDataAvailableAndFree():
            return

        PatientWrapper.setAndPublishBusyStatus(client, 1)

        MenuCLI.displayEditMenu()
        
        print("Enter your choice: ")
        choice = settings.get_user_choice()

        settings.edit_functions_map[choice]()

        settings.temporary_patient.setBusyStatus(0)
        msg = settings.temporary_patient.toString()
        client.publish(settings.topic, msg)


    @staticmethod
    def printPatientInformation():
        print("Name: ", settings.original_patient.getName())
        print("Age: ", settings.original_patient.getAge())
        print("Gender:", settings.original_patient.getGender())
        print("Procedure Name: ", settings.original_patient.getProcedureName())
        print("Consumables: ", settings.original_patient.getConsumables())
        print("Report IDs", settings.original_patient.getReportIds())


    @staticmethod
    def printPatientCount():
        patient_count = settings.get_patient_count()
        print("Patient Count: ", patient_count)